controllers/cluster_status.go

```diff
@@ -365,6 +365,24 @@ func (r *ClusterReconciler) updateResourceStatus(
 	return nil
 }
 
+// removeConditionsWithInvalidReason will remove every condition which is not valid
+// anymore from the K8s API point-of-view
+func (r *ClusterReconciler) removeConditionsWithInvalidReason(ctx context.Context, cluster *apiv1.Cluster) error {
+	conditions := make([]metav1.Condition, 0, len(cluster.Status.Conditions))
+	for _, entry := range cluster.Status.Conditions {
+		if utils.IsConditionReasonValid(entry.Reason) {
+			conditions = append(conditions, entry)
+		}
+	}
+
+	if !reflect.DeepEqual(cluster.Status.Conditions, conditions) {
+		cluster.Status.Conditions = conditions
+		return r.Status().Update(ctx, cluster)
```

Comment:
if we call this `r.Status.Update(ctx, cluster)` without finishing this reconcile, this will cause
- a new worker will pick up the update event in the work queue and proceed the cluster
- this reconcile will continue to proceed the cluster

and when they call update, one of them will fail due to the conflict
```sh
the object has been modified; please apply your changes to the latest version and try again
```